packages/graphile-utils/src/makeWrapResolversPlugin.ts

```diff
@@ -27,19 +27,33 @@ interface ResolverWrapperRequirements {
   siblingColumns?: Array<{ column: string; alias: string }>;
 }
 
-interface ResolverWrapperRule {
+interface ResolverWrapperRule<
+  TSource extends unknown = any,
+  TContext extends unknown = any,
+  TArgs extends unknown = { [argName: string]: any }
+> {
   requires?: ResolverWrapperRequirements;
-  resolve?: ResolverWrapperFn;
+  resolve?: ResolverWrapperFn<TSource, TContext, TArgs>;
   // subscribe?: ResolverWrapperFn;
 }
 
-interface ResolverWrapperRules {
+interface ResolverWrapperRules<
+  TSource extends unknown = any,
+  TContext extends unknown = any,
+  TArgs extends unknown = { [argName: string]: any }
+> {
   [typeName: string]: {
-    [fieldName: string]: ResolverWrapperRule | ResolverWrapperFn;
+    [fieldName: string]:
+      | ResolverWrapperRule<TSource, TContext, TArgs>
+      | ResolverWrapperFn<TSource, TContext, TArgs>;
```

Comment:
I'm not sure this really makes sense. TContext will be the same for all resolvers, but TSource and TArgs will be field specific, so in this two dimensional map it wouldn't give a way to indicate TSource and TArgs for each individual field. e.g.

```graphql
type User {
  name(short: Boolean = false): String
  age(in: DurationEnum!): Int
}
type Post {
  digest: String
}
```

```js
const rules = {
  User: {
    name: {
      requires: { siblingColumns: [
        {column: 'first_name', alias: '$firstName'},
        {column: 'lastst_name', alias: '$lastName'}
      ]},
      resolver(
        soruce: {$firstName: string, $lastName: string},
        args: {short: boolean},
        context: TContext
      ) { ... }
    }
    
    age: {
      requires: { siblingColumns: [
        {column: 'date_of_birth', alias: '$dob'}
      ]},
      resolver(
        source: {$dob: Date},
        args: {in: DurationEnum},
        context: TContext
      ) { ... }
    },
  },
  Post: { ... }
}
```

Here we can see the args and source types are different for each field.

Reply:
Hmm... good point. I pushed some changes; for the inference to properly work it seems like we'd need to do some changes to `ResolverWrapperRules`. If we change `ResolverWrapperRules` to a type instead of an interface we can use `in keyof` to go through all the keys of the `ResolverWrapperRules` instance, and then use conditional types to infer the generic type parameters of `ResolverWrapperFn` (or `ResolverWrapperRule`). 

These changes aren't quite backwards compatible though...

Reply:
Yeah; they'll need to wait for V5 I'm afraid. Thanks anyway!